### PR TITLE
Updated financial aid review page frontend

### LINKED ISSUE #1045: Financial Aid: Review Page
Simple list of financial aid submissions, for review. 

image

# Acceptance Criteria:

- [ ] Remove "reject" button  /cc @roberthouse54 
- [ ] Rename "approve" checkmark with label "Save"  /cc @roberthouse54 
- [ ] Move "adjusted cost" column close to "Save" button
- [ ] Add a dropdown with a couple of predefined values that will allow reviewer to log a justification for adjusting the cost (e.g  "Docs not notarized" , "Insufficient Docs")
- [ ] Button for documents received
- [ ] Date calculated
- [ ] Date docs sent
- [ ] Radio button for displaying income in dollars or local currency (as calculated, not real time)
- [ ] Modified `PUT` requests for `FinancialAidActionView` to `PATCH` requests to better follow REST
- [ ] Changed `FinancialAidMailView` to use a serializer and to take a `FinancialAid` object instead of email recipient for security and auto-message-generation reasons
- [ ] Linked backend functionality for emailing, approving, and marking documents as received to the frontend
- [ ] UI selectively shows/hides sections and functionality depending on the status selected to view
- [ ] Updated labels and copy
- [ ] Search by name functionality
- [x] Make some of the columns sortable 
- [x] API endpoint for requests given a status (searchable, sortable and pageable)
- [x] simple search for user name
- [x] fields for each request include income of an applicant (as submitted in local currency and as dollars as calculated), date of calculation, document status
- [x] Limit API endpoints to staff in the appropriate program
- [x] API endpoint to get dollar values of each tier (for populating drop-down)

### LINKED ISSUE #1214: Financial Aid: display "date sent" in the Review Form

### PR DESCRIPTION
**Pull request by alvinsiu**
#### What are the relevant tickets?
Closes #1045, closes #1214

#### What's this PR do?
Updates the financial aid review page frontend and backend:
- Moved `FinancialAidStatus` and `FinancialAidJustification` to `financialaid/constants.py`
- Added `FinancialAid.justification`
- Temporarily removed `FinancialAidStatus.REJECTED` by commenting out code as necessary as per request
- Modified `PUT` requests for `FinancialAidActionView` to `PATCH` requests to better follow REST, updating `FinancialAidActionSerializer`, its validation, and tests accordingly
- Modified the UI for the review page as requested
- Changed `FinancialAidMailView` to use a serializer and to take a `FinancialAid` object instead of email recipient for security and auto-message-generation reasons
- Linked backend functionality for emailing, approving, and marking documents as received to the frontend
- UI selectively shows/hides sections and functionality depending on the status selected to view
- Updated labels and copy as per request
- Added search by name functionality

#### Any background context you want to provide?
Since everything was pretty integrated across different parts of the codebase, I didn't make separate issues for the smaller components for the sake of time, hence the larger PR.

Screenshots below in the comments.

#### What GIF best describes this PR or how it makes you feel?
image

### CODE REVIEW

**Comment by alvinsiu** [comment]
@Ferdi @pdpinch Here's the updated frontend page as you requested.  I still need to hook in the backend, but I wanted to have this up for you to have a look at.

**Comment by pdpinch** [comment]
- [ ] Remove "reject" button  /cc @roberthouse54 
- [ ] Rename "approve" checkmark with label "Save"  /cc @roberthouse54 
- [ ] Move "adjusted cost" column close to "Save" button
- [ ] Add a dropdown with a couple of predefined values that will allow reviewer to log a justification for adjusting the cost (e.g  "Docs not notarized" , "Insufficient Docs")

How hard will it be to update the predefined values for justification?

**Comment by alvinsiu** [comment]
Pretty simple - just updating a constants file.

**Comment by Ferdi** [comment]
@alvinsiu thanks !

Can we not add that top row for sorting and instead use the existing headers ? I understand it wont work for "location" but I dont think that's important.

**Comment by alvinsiu** [comment]
Sure we can make that happen!

**Comment by Ferdi** [comment]
What does "Search" box do ?   If it search on the name field, can we change the label to "Search by name" ?

**Comment by alvinsiu** [comment]
The search box doesn't do anything at the moment - it was just on the mock, so I put it in there as a placeholder.  Otherwise, I don't think we've ever defined what that functionality should look like.

**Comment by alvinsiu** [comment]
Updated screenshots:

image
image
image
image
image
image

**Comment by alvinsiu** [comment]
Quite a few changes in this PR - a lot of them small tweaks uncovered while linking the back and frontends, and some of them slightly more involved.  Here are the general changes made:

- Moved `FinancialAidStatus` and `FinancialAidJustification` to `financialaid/constants.py`
- Added `FinancialAid.justification`
- Temporarily removed `FinancialAidStatus.REJECTED` by commenting out code as necessary as per request
- Modified `PUT` requests for `FinancialAidActionView` to `PATCH` requests to better follow REST, updating `FinancialAidActionSerializer`, its validation, and tests accordingly
- Modified the UI for the review page as requested
- Changed `FinancialAidMailView` to use a serializer and to take a `FinancialAid` object instead of email recipient for security and auto-message-generation reasons
- Linked backend functionality for emailing, approving, and marking documents as received to the frontend
- UI selectively shows/hides sections and functionality depending on the status selected to view
- Updated labels and copy as per request
- Added search by name functionality

**Comment by noisecapella** [comment]
In the future please limit the amount of refactoring in these PRs unless it's related to the issue at hand. Sorry for being blunt but the financial aid APIs are a bit messy and I would rather do that kind of refactoring when we're not under major time pressure, so we can do it all at once and thoughtfully. It also increases the review workload when we're under a time crunch right now

File: financialaid/models.py [hunk 1]
```diff
@@ -138,7 +106,11 @@ class FinancialAid(TimestampedModel):
     tier_program = models.ForeignKey(TierProgram, null=False)
     status = models.CharField(
         null=False,
-        choices=[(status, status) for status in FinancialAidStatus.ALL_STATUSES],
+        choices=[
+            (status, status) for status in FinancialAidStatus.ALL_STATUSES
+            # The rejected status has been temporarily disabled as a valid status for FinancialAid
```
Comment (noisecapella) [inline financialaid/models.py]: Please revert this change unless it's necessary for the work in this PR. We should not be putting in 'temporary' changes, especially right before a release. We should remove `rejected` with a separate PR

File: financialaid/serializers.py [hunk 2]
```diff
@@ -110,53 +113,70 @@ def save(self):
...
-        try:
-            data["tier_program"] = TierProgram.objects.get(
-                id=data["tier_program_id"],
-                program_id=self.instance.tier_program.program_id,
-                current=True
-            )
-        except TierProgram.DoesNotExist:
-            raise ValidationError("Financial Aid Tier does not exist for this program.")
+        # Required field
+        if data.get("action", None) is None:
```
Comment (noisecapella) [inline financialaid/serializers.py]: `get` doesn't need to be passed `None`, it is the default

File: financialaid/templates/review_financial_aid.html [hunk 3]
```diff
@@ -1,116 +1,409 @@
...
+            displayMessage("Error: " + result.responseText + " on " + name + "'s financial aid application.", "danger");
+          }
+        });
+      }
+    }
+    {% endif %}
+    
+    
+    {% if selected_status == statuses.PENDING_MANUAL_APPROVAL %}
+    /**
```
Comment (noisecapella) [inline financialaid/templates/review_financial_aid.html]: Can you move all this JS to a separate file in `static/js`? We also don't need to define the functions conditionally, it would be better if they were always defined

**Comment by noisecapella** [comment]
How does the user set justification for pending-manual-approval users?

screenshot from 2016-10-05 11-21-13

**Comment by roberthouse54** [comment]
@noisecapella I don't think we need a confirmation for either of these.

**Comment by alvinsiu** [comment]
@noisecapella @roberthouse54 What if they accidentally click the wrong button for the wrong user?  Or they've accidentally tabbed over and hit enter?  It seems dangerous to me to not have a confirmation of some sort.

Thoughts?

**Comment by roberthouse54** [comment]
I guess I am ok with a confirmation dialog, but this will maybe get annoying to see every time. But you are probably right.

File: financialaid/serializers.py [hunk 4]
```diff
@@ -110,53 +113,70 @@ def save(self):
...
+            raise ValidationError({"action": "This field is required."})
+        # For approving and rejecting
+        if data["action"] in [FinancialAidStatus.APPROVED, FinancialAidStatus.REJECTED]:
+            # Required fields
+            if data.get("tier_program_id", None) is None:
+                raise ValidationError({"tier_program_id": "This field is required."})
+            if data.get("justification", None) is None:
+                raise ValidationError({"justification": "This field is required."})
```
Comment (alvinsiu) [inline financialaid/serializers.py]: These are used on the frontend.

The dicts vs strings is how DRF differentiates between field errors and non-field errors.

File: financialaid/serializers.py (hunk 2 above)
Comment (alvinsiu) [inline financialaid/serializers.py]: Will remove these instances

File: financialaid/templates/review_financial_aid.html (hunk 3 above)
Comment (alvinsiu) [inline financialaid/templates/review_financial_aid.html]: Will do!

File: financialaid/serializers.py (hunk 4 above)
Comment (noisecapella) [inline financialaid/serializers.py]: Ah, I didn't know that

File: financialaid/models.py (hunk 1 above)
Comment (alvinsiu) [inline financialaid/models.py]: It's not a "temporary" change per se - it's supposed to be removed for 10/15, but I remember during a standup last week @pdpinch said to leave the code in for `rejected`, but commented out.

But you're right, we should remove `rejected` with a separate PR - will revert changes and do that there.

**Comment by noisecapella** [comment]
When I search using the textbox I get this exception:

screenshot from 2016-10-05 17-17-21

**Comment by alvinsiu** [comment]
Ah, forgot to remove django tags when moving the javascript.

**Comment by noisecapella** [comment]
I'm getting an error when clicking 'mark docs received':

screenshot from 2016-10-06 09-59-55

It looks like there is django template markup in the JS file. This should be passed as a parameter instead

**Comment by alvinsiu** [comment]
I should have caught that in testing, my bad - I've re-tested all of the actions, so it should work as expected now.

File: static/js/financial_aid_review.js [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+  "use strict";
+  
+  var CSRF_TOKEN = window.CSRF_TOKEN;
+  var BASE_PATH = window.BASE_PATH;
+  
+  /**
+   * Marks documents as received for a financial aid application
+   * 
+   * @param financialAidId {number} Financial aid application id
+   * @param url {string} URL to submit request to
```
Comment (noisecapella) [inline static/js/financial_aid_review.js]: Can you add `action` here?

File: static/js/financial_aid_review.js [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        }
+      });
+    }
+  }
+  
+  /**
+   * Submits a financial aid application approval
+   * 
+   * @param financialAidId {number} Financial aid application id
+   * @param url {string} URL to submit request to
```
Comment (noisecapella) [inline static/js/financial_aid_review.js]: Here as well

File: static/js/financial_aid_review.js (hunk 5 above)
Comment (alvinsiu) [inline static/js/financial_aid_review.js]: Yup!

File: static/js/financial_aid_review.js (hunk 6 above)
Comment (alvinsiu) [inline static/js/financial_aid_review.js]: Done!

**Comment by noisecapella** [comment]
Can you download the jquery and bootstrap files that you're using and put them in `static/js/financial_aid`? Also please move the `financial_aid_review.js` there too

**Comment by alvinsiu** [comment]
Sure - I'll do that once I finish running tests for merging with master!

File: financialaid/views.py [hunk 7]
```diff
@@ -164,43 +172,36 @@ def get_context_data(self, **kwargs):
         ).annotate(
```
Comment (noisecapella) [inline financialaid/views.py]: We need to filter on `current=True` here

Comment (alvinsiu) [inline financialaid/views.py]: Good catch!

File: financialaid/templates/review_financial_aid.html [hunk 8]
```diff
@@ -1,116 +1,276 @@
...
               </td>
-              <td>${{ financial_aid.income_usd }}</td>
-              <td>{{ financial_aid.created_on|date:"SHORT_DATE_FORMAT" }}</td>
+              <!-- /Adjusted cost -->
+              <!-- Justification -->
+              {% if selected_status == statuses.PENDING_MANUAL_APPROVAL %}
               <td>
-                <button class="btn btn-default">
-                  <span class="glyphicon glyphicon-ok" style="color: #0074e1"></span>
+                {% if selected_status == statuses.PENDING_MANUAL_APPROVAL %}
```
Comment (noisecapella) [inline financialaid/templates/review_financial_aid.html]: Can you remove this if statement? It seems redundant

Comment (alvinsiu) [inline financialaid/templates/review_financial_aid.html]: Ah, this was intentional, but I messed up fixing an earlier bug.  This is to show the justification for approved applications, but render the `<select>` if it's pending manual approval.  I'll make the change.

File: financialaid/views_test.py [hunk 9]
```diff
@@ -252,25 +256,52 @@ def test_review_financial_aid_view_with_invalid_filter_and_sorting(self):
...
+            path=self.review_url_with_filter,
+            search_query=search_query
+        )
+        resp = self.assert_http_status(self.client.get, search_url, status.HTTP_200_OK)
+        resp_obj_id_list = resp.context_data["financial_aid_objects"].values_list("id", flat=True)
+        expected_obj_id_list = FinancialAid.objects.filter(
+            Q(user__profile__first_name__icontains=search_query) | Q(user__profile__last_name__icontains=search_query),
+            tier_program__program_id=self.program.id,
+            status=FinancialAidStatus.AUTO_APPROVED
+        ).order_by("user__profile__first_name").values_list("id", flat=True)  # Default sort field
```
Comment (noisecapella) [inline financialaid/views_test.py]: The default sort is `last_name`.

By the way, one of the problems with `UserFactory` is that it doesn't generate a profile using `ProfileFactory`, a blank profile is created automatically with a `post_save` signal. I think this works only because all profiles here are blank. Can you change instances of `FinancialAidFactory` to use profiles generated using `ProfileFactory`, maybe with a helper function?

File: financialaid/serializers.py [hunk 10]
```diff
@@ -119,37 +122,52 @@ class FinancialAidActionSerializer(serializers.Serializer):
...
-            data["tier_program"] = TierProgram.objects.get(
-                id=data["tier_program_id"],
-                program_id=self.instance.tier_program.program_id,
-                current=True
-            )
-        except TierProgram.DoesNotExist:
-            raise ValidationError("Financial Aid Tier does not exist for this program.")
+        # Required field
+        if data.get("action") is None:
+            raise ValidationError({"action": "This field is required."})
```
Comment (noisecapella) [inline financialaid/serializers.py]: Can you add tests to cover this line and the two other `This field is required` validation errors below?

File: financialaid/views_test.py (hunk 9 above)
Comment (alvinsiu) [inline financialaid/views_test.py]: Ah, good catch.

And yup, I can make changes to the FinancialAidFactory.

Comment (alvinsiu) [inline financialaid/views_test.py]: I made this change by overriding `FinancialAidFactory.create()` so that it retains the same workflow as every other factory.  It's basically the same thing as what you suggested, just not as a separate function to call.

Let me know if the comments I have on it aren't clear enough or if you think of any issues with this solution.

Comment (noisecapella) [inline financialaid/views_test.py]: Looks fine to me

File: financialaid/views_test.py [hunk 11]
```diff
@@ -204,13 +208,13 @@ def test_review_financial_aid_view_with_filter_and_sorting(self):
         FinancialAidFactory.create(tier_program=self.tier_programs["0k"], status=FinancialAidStatus.AUTO_APPROVED)
```
Comment (noisecapella) [inline financialaid/views_test.py]: Can we add more `FinancialAid`s with the same status? With one element for each we can't verify sort

File: financialaid/views_test.py [hunk 12]
```diff
@@ -252,25 +256,52 @@ def test_review_financial_aid_view_with_invalid_filter_and_sorting(self):
         url_with_bad_filter_and_bad_sorting = "{url}?sort_by=-askjdf".format(url=url_with_bad_filter)
         self.assert_http_status(self.client.get, url_with_bad_filter_and_bad_sorting, status.HTTP_200_OK)
 
+    def test_review_financial_aid_view_with_search(self):
+        """
+        Tests that ReviewFinancialAidView returns the expected results with search
+        """
+        FinancialAidFactory.create(tier_program=self.tier_programs["0k"], status=FinancialAidStatus.AUTO_APPROVED)
```
Comment (noisecapella) [inline financialaid/views_test.py]: Can you add more copies here too?

**Comment by noisecapella** [comment]
It looks like there's a missing migration due to the status order change

File: financialaid/views_test.py (hunk 11 above)
Comment (alvinsiu) [inline financialaid/views_test.py]: Sounds good!

**Comment by noisecapella** [comment]
Can you rebase please?

File: static/js/financial_aid/bootstrap.min.js [hunk 13]
```diff
@@ -0,0 +1,8 @@
+/* eslint-disable */
```
Comment (alvinsiu) [inline static/js/financial_aid/bootstrap.min.js]: Yup!

**Comment by alvinsiu** [comment]
Wow, just rebased an hour ago and already conflicts :/

**Comment by noisecapella** [comment]
Ok I think that's everything :+1:

**Comment by alvinsiu** [comment]
Thanks for the really thorough review on this really long PR!
